{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import requests\n",
    "from Bio import SeqIO\n",
    "from functools import lru_cache\n",
    "url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils'\n",
    "\n",
    "def read_fasta(file, raw=False) -> dict:\n",
    "    \"\"\"\n",
    "    Reads information from a fasta formatted file.\n",
    "    \"\"\"\n",
    "    seq_raw = SeqIO.parse(file, \"fasta\")\n",
    "    if raw : return seq_raw\n",
    "    seq_dict = {}\n",
    "    for seq in seq_raw:\n",
    "        raw_id = seq.id.split(\"|\")\n",
    "        if len(raw_id) > 2:\n",
    "            id = raw_id[3]\n",
    "            seq.id = raw_id[3]\n",
    "        elif len(raw_id) == 2:\n",
    "            if raw_id[0] == raw_id[1]: id = raw_id[0]\n",
    "            else: id = seq.id\n",
    "        else:\n",
    "            id = seq.id\n",
    "        seq_dict[id] = seq\n",
    "    return seq_dict\n",
    "\n",
    "def get_ncbi(id: str):\n",
    "    params = {\n",
    "        'db': 'nucleotide',\n",
    "        'id': id,\n",
    "        'retmode': 'text',\n",
    "        'rettype': 'text'\n",
    "    }\n",
    "    r = requests.get(f'{url}/efetch.fcgi', params=params)\n",
    "    try:\n",
    "        match = re.search(r'strain\\s(.[^\\s]+)|strain=\"(.[^\\s]+)\"|subtype:\\s(.[^\\s;\"]*)|(subtype\\s[A-Z])', r.text)\n",
    "        result = \"\".join(match.groups(\"\"))\n",
    "        if 'subtype' not in result:\n",
    "            if 'CRF' not in result: \n",
    "                result = 'subtype ' + result\n",
    "        return result\n",
    "    except AttributeError:\n",
    "        if 'NL4-3' in r.text: return 'NL4-3'\n",
    "        elif 'HXB2' in r.text: return 'HXB2'\n",
    "        else: return None\n",
    "\n",
    "def serch_hivdb(seq):\n",
    "    payload = {\n",
    "        'operationName': 'guess_type',\n",
    "        'query': 'query example($sequences: [UnalignedSequenceInput]!) {currentVersion { text, publishDate },sequenceAnalysis(sequences: $sequences) {bestMatchingSubtype { displayWithoutDistance, distance },}}',\n",
    "        'variables': {\n",
    "            'sequences': [{'header': id, 'sequence': str(seq)}]\n",
    "        }\n",
    "    }\n",
    "    hivdb = requests.post('https://hivdb.stanford.edu/graphql', json=payload)\n",
    "    best_match = hivdb.json().get('data').get('sequenceAnalysis')[0].get('bestMatchingSubtype')\n",
    "    subtype = 'subtype ' + best_match.get('displayWithoutDistance')\n",
    "    conf = (1 - float(best_match.get('distance')))*100\n",
    "    return f'{subtype} ({conf}%)'\n",
    "     \n",
    "@lru_cache(maxsize=128)\n",
    "def get_subtype(id: str, fasta=None):\n",
    "    result = get_ncbi(id)\n",
    "    if result: return result\n",
    "    if fasta: return serch_hivdb(fasta)\n",
    "    return 'UNKNOWN'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "id = 'X01762.1'\n",
    "params = {\n",
    "    'db': 'nucleotide',\n",
    "    'id': id,\n",
    "    'retmode': 'text',\n",
    "    'rettype': 'text'\n",
    "}\n",
    "r = requests.get(f'{url}/efetch.fcgi', params=params)\n",
    "try:\n",
    "    match = re.search(r'strain\\s(.[^\\s]+)|strain=\"(.[^\\s]+)\"|subtype:\\s(.[^\\s;\"]*)|(subtype\\s[A-Z])', r.text)\n",
    "    print(''.join(match.groups('')))\n",
    "except AttributeError:\n",
    "    if 'NL4-3' in r.text: \n",
    "        print(f'{id} - NL4-3')\n",
    "    else:\n",
    "        payload = {\n",
    "            'operationName': 'guess_type',\n",
    "            'query': 'query example($sequences: [UnalignedSequenceInput]!) {currentVersion { text, publishDate },sequenceAnalysis(sequences: $sequences) {bestMatchingSubtype { display },}}',\n",
    "            'variables': {\n",
    "                'sequences': [{'header': id, 'sequence': str(fa[id].seq)}]\n",
    "            }\n",
    "        }\n",
    "        hivdb = requests.post('https://hivdb.stanford.edu/graphql', json=payload)\n",
    "        try:\n",
    "            subtype = hivdb.json().get('data').get('sequenceAnalysis')[0].get('bestMatchingSubtype').get('display')[0]\n",
    "            print(f'{id} - subtype {subtype} (guessed)')\n",
    "        except:\n",
    "            print(hivdb.text)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded: 32 sequence(s)\n",
      "AF164485.1 - subtype E\n",
      "AF259954.1 - subtype E\n",
      "AF259955.1 - subtype E\n",
      "KU168309.1 - CRF01_AE\n",
      "U51189.1 - subtype E\n",
      "AB049811.1 - subtype A\n",
      "AB231895.1 - subtype G\n",
      "KC492737 - CRF07_BC\n",
      "KC492738 - CRF07_BC\n",
      "KC503852 - CRF07_BC\n",
      "KC503853 - CRF07_BC\n",
      "KC503854 - CRF07_BC\n",
      "KC503855 - CRF07_BC\n",
      "KF234628 - CRF07_BC\n",
      "AF385934 - CRF12_BF\n",
      "AF385935 - CRF12_BF\n",
      "AF385936 - CRF12_BF\n",
      "AB703607.1 - CRF35_AD\n",
      "EF158040.1 - CRF35_AD\n",
      "AF324493.2 - NL4-3\n",
      "MH705157.1 - subtype A\n",
      "AY835759.1 - subtype B\n",
      "K03455.1 - HXB2\n",
      "X01762.1 - subtype B (99.22562478000704%)\n",
      "AY772699.1 - subtype C\n",
      "AB485648.1 - subtype D (94.61457233368532%)\n",
      "AJ320484.1 - subtype 92UG001.\n",
      "KY392769.1 - subtype D (100.0%)\n",
      "AB231893.1 - subtype G\n",
      "AB485662.1 - subtype G (100.0%)\n",
      "AF084936.1 - subtype G\n",
      "MH705134.1 - subtype G\n"
     ]
    }
   ],
   "source": [
    "fa = read_fasta('O:\\\\OneDrive\\\\Projects\\\\Project-HIV_analysis_pipeline\\\\workflow\\\\references\\\\HIV1_Genomes_32Strains.fasta')\n",
    "print(f'Loaded: {len(fa.keys())} sequence(s)')\n",
    "for id in fa.keys():\n",
    "    print(f'{id} - {get_subtype(id, fa[id].seq)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AF164485.1 - subtype CRF01_AE (98.38085181274198%)\n",
      "AF259954.1 - subtype CRF01_AE (100.0%)\n",
      "AF259955.1 - subtype CRF01_AE (100.0%)\n",
      "KU168309.1 - subtype CRF01_AE (97.571277719113%)\n",
      "U51189.1 - subtype CRF01_AE (98.80323829637452%)\n",
      "AB049811.1 - subtype CRF02_AG (100.0%)\n",
      "AB231895.1 - subtype CRF02_AG (99.82400563181977%)\n",
      "KC492737 - subtype CRF07_BC (97.9936642027455%)\n",
      "KC492738 - subtype CRF07_BC (97.95846532910947%)\n",
      "KC503852 - subtype CRF07_BC (97.78247096092925%)\n",
      "KC503853 - subtype CRF07_BC (97.7472720872932%)\n",
      "KC503854 - subtype CRF07_BC (97.85286870820133%)\n",
      "KC503855 - subtype CRF07_BC (97.8176698345653%)\n",
      "KF234628 - subtype CRF07_BC (98.13445969728969%)\n",
      "AF385934 - subtype CRF12_BF (100.0%)\n",
      "AF385935 - subtype CRF12_BF (96.83210137275607%)\n",
      "AF385936 - subtype CRF12_BF (96.58570925730376%)\n",
      "AB703607.1 - subtype CRF35_AD (100.0%)\n",
      "EF158040.1 - subtype CRF35_AD (97.92326645547342%)\n",
      "AF324493.2 - subtype B (98.20485744456178%)\n",
      "MH705157.1 - subtype A (95.91693065821893%)\n",
      "AY835759.1 - subtype B (98.59204505455826%)\n",
      "K03455.1 - subtype B (99.29602252727912%)\n",
      "X01762.1 - subtype B (99.22562478000704%)\n",
      "AY772699.1 - subtype C (96.30411826821542%)\n",
      "AB485648.1 - subtype D (94.61457233368532%)\n",
      "AJ320484.1 - subtype D (100.0%)\n",
      "KY392769.1 - subtype D (100.0%)\n",
      "AB231893.1 - subtype G (99.64801126363957%)\n",
      "AB485662.1 - subtype G (100.0%)\n",
      "AF084936.1 - subtype G (99.92960225272792%)\n",
      "MH705134.1 - subtype G (96.97289686730025%)\n"
     ]
    }
   ],
   "source": [
    "for id in fa.keys():\n",
    "    print(f'{id} - {serch_hivdb(fa[id].seq)}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
